common/src/main/java/io/druid/collections/ReferenceCountingResourceHolder.java

```diff
@@ -0,0 +1,135 @@
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied. See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+package io.druid.collections;
+
+import com.metamx.common.ISE;
+import com.metamx.common.logger.Logger;
+
+import java.io.Closeable;
+import java.util.concurrent.atomic.AtomicBoolean;
+
+public class ReferenceCountingResourceHolder<T> implements ResourceHolder<T>
+{
+  private static final Logger log = new Logger(ReferenceCountingResourceHolder.class);
+
+  private final Object lock = new Object();
+  private final T object;
+  private final Closeable closer;
+
+  private long refcount = 1;
+  private boolean didClose = false;
+
+  public ReferenceCountingResourceHolder(final T object, final Closeable closer)
+  {
+    this.object = object;
+    this.closer = closer;
+  }
+
+  @Override
+  public T get()
+  {
+    synchronized (lock) {
+      if (refcount <= 0) {
+        throw new ISE("Already closed!");
+      }
+
+      return object;
+    }
+  }
+
+  public ReferenceCountingResourceHolder.Releaser increment()
+  {
+    synchronized (lock) {
+      if (refcount <= 0) {
+        throw new ISE("Already closed!");
+      }
+
+      refcount++;
+
+      final AtomicBoolean didRelease = new AtomicBoolean();
+      return new Releaser()
+      {
+        @Override
+        public void release()
+        {
+          if (didRelease.compareAndSet(false, true)) {
+            decrement();
+          } else {
+            log.warn("WTF?! release called but we are already released!");
+          }
+        }
+
+        @Override
+        protected void finalize() throws Throwable
+        {
+          if (didRelease.compareAndSet(false, true)) {
+            log.warn("Not released! Object was[%s], releasing on finalize of releaser.", object);
+            decrement();
+          }
+        }
+      };
+    }
+  }
+
+  @Override
+  public void close()
+  {
+    synchronized (lock) {
+      if (!didClose) {
+        didClose = true;
+        decrement();
+      } else {
+        log.warn(new ISE("Already closed!"), "Already closed");
+      }
+    }
+  }
+
+  @Override
+  protected void finalize() throws Throwable
+  {
+    synchronized (lock) {
+      if (!didClose) {
+        log.warn("Not closed! Object was[%s], closing on finalize of holder.", object);
+        didClose = true;
+        decrement();
+      }
+    }
+  }
+
+  private void decrement()
+  {
+    synchronized (lock) {
+      refcount--;
+      if (refcount <= 0) {
+        try {
+          closer.close();
+        }
+        catch (Exception e) {
+          log.error(e, "WTF?! Close failed, uh oh...");
+        }
+      }
+    }
+  }
+
+  public interface Releaser
```

Comment:
Hmm, it can extend Closeable though (and drop the throws clause). I'll change that.